### ISSUE TITLE
Marshal tests delegating to JSON.parse may result in unexpected SyntaxError message strings

### ISSUE THREAD

**Issue by SMotaal** [endojs/endo#1414]
The following test will fail on newer versions of Node.js and likely other runtimes:

https://github.com/endojs/endo/blob/dbd498e30a5c3b0d2713d863bc7479ceef39cd79/packages/marshal/test/test-marshal-stringify.js#L57-L59

I would like to get some input from folks before drafting a PR from my end to reduce the friction of having to rely downgrade just to ensure this test passes.

---

#### Considerations

- It is not clear if it is the intent that marshal would delegate to `JSON.parse` for this specific case:
  - This was the only test in the `marshal parse errors` which failed on my end.
  - There are other tests that look for `{message}` strings that are obviously specific to marshal and not what `JSON.parse` itself would report.
- It it is the intent:
  - The error thrown by [`JSON.parse` per the ECMAScript specification](https://tc39.es/ecma262/#sec-json.parse) must be of the type `SyntexError`.
  - This would [explicitly determine its `{name}`](https://tc39.es/ecma262/#sec-nativeerror.prototype.name).
  - This would not determine the `{message}` which is left to the discretion of implementers.
  - This issue may also be relevant in other places as it relates to changes made recently to "improve error messages" in https://github.com/v8/v8/commit/718f743750135b3a3e643655a0406a90bf4bf010.

#### Alternatives

1. Keep revising matcher to make it work for Node.js and other specific platforms moving forward:
   ```js
     t.throws(() => parse('X'), {
       message: /\b[Un]expected\b/,
     });
   ```
2. Switch from `{message}` to `{name}` checks:
   ```js
     t.throws(() => parse('X'), {
       name: /^SyntaxError$/,
     });
   ```
3. ~~Revise the test to dynamically prepare the actual and expected values.~~ (counterintuitive)

**Comment by mhofman** [comment]
Yeah it seems that checking for a `.name` to match `SyntaxError` would be the correct test to have. @mfig @erights WDYT ?

**Comment by michaelfig** [comment]
Another alternative to consider:
4. Use `e instanceof SyntaxError`:
```js
t.throws(() => parse('X'), {
  instanceOf: SyntaxError,  
});
```

I'm unsure of the relative merits of `name:` vs `instanceOf:`, but either would be fine with me!

**Comment by erights** [comment]
Huh. I didn't know that Ava had all these options. Neat.

I mildly prefer `instanceof:` to `name:`, since it is less likely to give a false signal if something is screwed up, like the error object having its own `name` property that disagrees with its kind.

**Comment by SMotaal** [comment]
I have a couple of questions for clarity:

- Is it worth considering to normalize the JSON.parse error messages across runtimes internally within Marshal itself?

- Are there any possibilities the instance would belong to a different `SyntaxError` constructor?

I am not up-to-date to know for sure if identity discontinuity could be a problem in some scenario I am missing, although I confirmed that the [ShadowRealms proposal at stage 3](https://github.com/tc39/proposal-shadowrealm/blob/main/explainer.md) is definitive about it not being so.

**Comment by erights** [comment]
Scope: My preference is to fix everywhere in the monorepo. Al least all the obvious cases. Any others should at least be annotated with a TODO comment

Consistency: These are errors determined by the underlying implementation, not by our code. The spec never specifies the contents of error messages, only the type of error. We should generally not test behavior not determined by the spec, unless that issue specifically is part of the purpose of that test. If you find such cases, please at least leave behind an explanatory comment.

Conformance: Good point! Given that, I much prefer testing `instanceof` to `name` since it will detect any identity discontinuity surprises. (Not that I'm expecting any)

Thanks!

**Comment by SMotaal** [comment]
Outside the scope of this issue and the PR, I would note that I am not seeing any errors in Safari:

```js
Object.getPrototypeOf((0, eval)('(async function() {})')).constructor('')

// async function anonymous() {
// 
// }
```

I only bring this up because I am not sure if this would lead to failures on mobile platforms where browsers delegate to WebViews and if that would relevant to investigate or address elsewhere.
